libs/hmpb/src/render_ballot.test.tsx

```diff
@@ -0,0 +1,66 @@
+import { electionFamousNames2021Fixtures } from '@votingworks/fixtures';
+import { test, expect } from 'vitest';
+import { BallotType } from '@votingworks/types';
+import {
+  allBaseBallotProps,
+  BaseBallotProps,
+  renderMinimalBallotsToCreateElectionDefinition,
+} from './render_ballot';
+import { BALLOT_MODES } from './types';
+import { createPlaywrightRenderer } from './playwright_renderer';
+import { ballotTemplates } from './ballot_templates';
+
+function combinations<T extends Record<string, unknown>>(
+  arrays: Array<Array<Partial<T>>>
+): T[] {
+  return arrays.reduce(
+    (acc, array) =>
+      acc.flatMap((accItem) =>
+        array.map((arrayItem) => ({ ...accItem, ...arrayItem }))
+      ),
+    [{}]
+  ) as T[];
+}
+
+test('allBaseBallotProps creates props all possible ballots for an election', () => {
+  const election = electionFamousNames2021Fixtures.readElection();
+  const allBallotProps = allBaseBallotProps(election);
+  const expectedPropCombos = combinations<
+    Pick<
+      BaseBallotProps,
+      'ballotStyleId' | 'precinctId' | 'ballotType' | 'ballotMode'
+    >
+  >([
+    election.ballotStyles.flatMap((ballotStyle) =>
+      ballotStyle.precincts.map((precinctId) => ({
+        ballotStyleId: ballotStyle.id,
+        precinctId,
+      }))
+    ),
+    [{ ballotType: BallotType.Absentee }, { ballotType: BallotType.Precinct }],
+    BALLOT_MODES.map((ballotMode) => ({ ballotMode })),
+  ]);
+  expect(allBallotProps).toHaveLength(expectedPropCombos.length);
+  for (const expectedPropCombo of expectedPropCombos) {
+    const expectedProps: BaseBallotProps = { ...expectedPropCombo, election };
+    expect(allBallotProps).toContainEqual(expectedProps);
+  }
+  for (const actualProps of allBallotProps) {
+    expect(actualProps.watermark).toBeUndefined();
+  }
```

Comment:
I love the coverage but I usually have 2 reservations with completely generated tests like this:

1. Makes it harder for readers to understand what is being tested (and thus update)
2. Increases risk of the test drifting in the future since generation of test case input and expectations are untested

I think this test is great for the situation because a large part of `allBaseBallotProps` is producing ballot props for all combinations of inputs, but could we add a second test for a single combination of `'ballotStyleId' | 'precinctId' | 'ballotType' | 'ballotMode'`? That would at least address concern #1 directly.

Reply:
Added: 7e3a145091d5d3678979e9d94a208a9051edf937